tagstudio/src/qt/ts_qt.py

```diff
@@ -492,6 +504,47 @@ def show_error_message(self, message: str):
         # Show the message box
         msg_box.exec()
 
+    def update_menu_actions(self):
+        # To override the enabling and disabling of buttons
+        # there are 2 data options that can be added to a button
+        # but are not necessary
+
+        # name_of_action.setData({"enable":False})
+
+        # this one should be added to buttons that do not need to be enabled when a library
+        # is opened for example the autofil button only enables
+        # when clicked on a item not on library open
+
+        # name_of_action.setData({"disable":False})
+        # this one should be added to a button that needs to be enabled
+        # even when not in a library like the "open library button"
+
+        # if the button should be controlled here then just do not add any data objects
+
+        if self.lib.library_dir:  # i think this should not return true when in a library?
+            # we are in a library! time to enable them.
+            for menu in self.menus:
+                for action in menu.actions():
+                    # scary but all this dose is check if the aforementioned "enable" option is
+                    # set to false and if it is then do not enable the button
+                    action.setDisabled(
+                        False
+                        if action.property("enable") is None
+                        else not action.property("enable")
+                        if action.property("enable") is False
+                        else False
+                    )
```

Comment:
At this point I feel the amount of crisscrossing boolean logic is simply a [trap](https://adamj.eu/tech/2021/07/10/python-type-hints-how-to-avoid-the-boolean-trap/) waiting to confuse anyone in the future who will need to touch this code.

Let's take a step back and look at what we'd like to do:
1. There's several menu options that we'd like to always *enable* whenever a library is open
2. There's several menu options that we'd always like to *disable* whenever a library is *not* open
3. There's a few menu options we'd like to only disable under specific library circumstances (i.e. there's nothing in the user's clipboard to "Paste" with

The use of `setProperty`/`setData` here is probably unnecessary to accomplish this. Instead we can simply add references to each menu item we know should always be disabled when a library is closed to a `list`. When a library is closed, we can call logic to iterate over that list and ensure that all options are disabled, regardless of their previous state. This list could be called something clear, such as "`menus_to_disable_on_close`". Likewise we could have another list that only contains references to menus that we know we *always* want open when a library is open. This would be iterated over in a similar fashion, this time enabling the options.

For any other cases, such as the "Paste" example mentioned earlier, we leave that logic up to the library. And if we were to have cases where an option should always be disabled (such as an incomplete feature, etc.) then we can just disable that on startup.

These lists would have clear names that let other people reading the code know what they're for. In the update method(s), the loops would be clearly iterating over them to perform the same unambiguous state change. If we ever find ourselves needing to account for a third predictable scenario, say if the program gains an additional "view mode" that changes what menu options are available, then we can easily follow the pattern we've laid out and add one or more additional lists of menu item references that we can cleanly iterate over during update methods. This ensures that developers can quickly see and understand the pattern that is here, and have an extensible pattern to work off of to build future features.